### PR TITLE
fix: chain halt if some invalid asset lock transactions are in mempool

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
As discovered during platform testing by @shumkov , it seems as the chain can halt in miner if somehow  mempool would have several transactions that are somehow invalid (maybe too low fee or something else). They can't be mined, but miner can't prepare a valid block with correct Credit Pool amount.

It is indeed can happen although I haven't reproduced it with functional tests at the moment 🤷‍♂️

## What was done?
Refactored and simplified a logic of Credit Pool amount of validation and added one more layer of validation: after all transaction are actually added to block by miner, it is recalculated one more time.
Also used correct `pindexPrev` instead Tip() for EHF signals.

## How Has This Been Tested?
Before this changes platform failed with this error and chain halt:
```
2023-10-20T06:20:16Z (mocktime: 2023-10-20T06:28:29Z) ERROR: ConnectBlock(DASH): CheckCreditPoolDiffForBlock for block 9d635e1fd0d7a8a5bf16ce158d3a39cbf903864bb6d671769836ea7db6055230 failed with bad-cbtx-asse locked-amount
```
With changes from this PR platform is generate the asset-lock transactions that are included to block and chain is not halt:
```
2023-10-27T10:45:37Z (mocktime: 2023-10-27T14:37:22Z) GetCreditPoolDiffForBlock: CCreditPool is CCreditPool(locked=32100015, currentLimit=32100015)
```

unit/functional tests are succeed.

## Breaking Changes
N/A; no consensus rules are changed

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by UdjinM6 — DISMISSED** [review]
utACK

File: src/evo/creditpool.cpp [hunk 1]
```diff
@@ -321,3 +293,30 @@ bool CCreditPoolDiff::ProcessLockUnlockTransaction(const CTransaction& tx, TxVal
         return state.Invalid(TxValidationResult::TX_CONSENSUS, "failed-procassetlocksinblock");
     }
 }
+
+std::optional<CCreditPoolDiff> GetCreditPoolDiffForBlock(const CBlock& block, const CBlockIndex* pindexPrev, const Consensus::Params& consensusParams,
+                                                         const CAmount blockSubsidy, BlockValidationState& state)
+{
+    AssertLockHeld(cs_main);
```
Comment (PastaPastaPasta) [inline src/evo/creditpool.cpp:300]: why does this need cs_main?

Comment (knst) [reply src/evo/creditpool.cpp]: removed. Seems just inherited from similar code around, I don't see a reason why it is needed.

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Review by ogabrielides — DISMISSED** [review]
utACK

File: src/evo/creditpool.cpp [hunk 2]
```diff
@@ -215,37 +215,18 @@ CCreditPoolManager::CCreditPoolManager(CEvoDB& _evoDb)
...
-}
 
-bool CCreditPoolDiff::SetTarget(const CTransaction& tx, const CAmount blockSubsidy, TxValidationState& state)
-{
-    CCbTx cbTx;
-    if (!GetTxPayload(tx, cbTx)) {
-        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-cbtx-payload");
+    if (llmq::utils::IsMNRewardReallocationActive(pindex)) {
+        // We consider V20 active if mn_rr is active
+        platformReward = MasternodePayments::PlatformShare(GetMasternodePayment(pindex->nHeight, blockSubsidy, /* v20_active= */ true));
```
Comment (PastaPastaPasta) [inline src/evo/creditpool.cpp:227]:
```suggestion
        platformReward = MasternodePayments::PlatformShare(GetMasternodePayment(pindex->nHeight, blockSubsidy, /*fV20Active=*/ true));
```

File: src/evo/creditpool.cpp [hunk 3]
```diff
@@ -215,37 +215,18 @@ CCreditPoolManager::CCreditPoolManager(CEvoDB& _evoDb)
...
+        assert(MoneyRange(platformReward));
```
Comment (PastaPastaPasta) [inline src/evo/creditpool.cpp:228]: this is duplicative; PlatformShare already asserts money range

Comment (UdjinM6) [reply src/evo/creditpool.cpp:228]:
```suggestion
```

File: src/evo/creditpool.cpp [hunk 4]
```diff
@@ -215,37 +215,17 @@ CCreditPoolManager::CCreditPoolManager(CEvoDB& _evoDb)
...
-}
 
-bool CCreditPoolDiff::SetTarget(const CTransaction& tx, const CAmount blockSubsidy, TxValidationState& state)
-{
-    CCbTx cbTx;
-    if (!GetTxPayload(tx, cbTx)) {
-        return state.Invalid(TxValidationResult::TX_CONSENSUS, "bad-cbtx-payload");
+    if (llmq::utils::IsMNRewardReallocationActive(pindex)) {
+        // We consider V20 active if mn_rr is active
+        platformReward = MasternodePayments::PlatformShare(GetMasternodePayment(pindex->nHeight, blockSubsidy, /*v20_active=*/ true));
```
Comment (UdjinM6) [inline src/evo/creditpool.cpp:227]: @knst note the name of the param, not just whitespaces
```suggestion
        platformReward = MasternodePayments::PlatformShare(GetMasternodePayment(pindex->nHeight, blockSubsidy, /*fV20Active=*/ true));
```

File: src/evo/creditpool.cpp (hunk 2 above)
Comment (knst) [reply src/evo/creditpool.cpp]: @PastaPastaPasta  maybe we can have linter for it? although I am not sure how to make it

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

File: src/miner.cpp [hunk 5]
```diff
@@ -403,10 +393,24 @@ void BlockAssembler::SortForBlock(const CTxMemPool::setEntries& package, std::ve
...
+void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpdated, const CBlockIndex* const pindexPrev)
 {
+    AssertLockHeld(cs_main); // for GetMNHFSignalsStage()
     AssertLockHeld(m_mempool.cs);
 
+    // This credit pool is used only to check withdrawal limits and to find
+    // duplicates of indexes. There's used `BlockSubsidy` equaled to 0
+    std::optional<CCreditPoolDiff> creditPoolDiff;
+    bool fV20Active_context = llmq::utils::IsV20Active(pindexPrev);
+    if (fV20Active_context) {
```
Comment (PastaPastaPasta) [inline src/miner.cpp:405]:
```suggestion
    if (llmq::utils::IsV20Active(pindexPrev)) {
```

File: src/evo/creditpool.cpp [hunk 6]
```diff
@@ -215,37 +215,17 @@ CCreditPoolManager::CCreditPoolManager(CEvoDB& _evoDb)
 {
 }
 
-CCreditPoolDiff::CCreditPoolDiff(CCreditPool starter, const CBlockIndex *pindex, const Consensus::Params& consensusParams) :
+CCreditPoolDiff::CCreditPoolDiff(CCreditPool starter, const CBlockIndex *pindex, const Consensus::Params& consensusParams, const CAmount blockSubsidy) :
```
Comment (PastaPastaPasta) [inline src/evo/creditpool.cpp:218]: consideration for future refactoring; I think we should make `starter` a forwarding reference

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Review by UdjinM6 — APPROVED** [review]
rererere-utACK :D

**Pull request merged**